File: java-example/src/main/java/demo/actors/HealthCheckActor.java
package demo.actors;

import java.time.Duration;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

import akka.actor.AbstractActor;
import akka.actor.ActorRef;
import akka.actor.Cancellable;
import akka.actor.Props;
import akka.event.Logging;
import akka.event.LoggingAdapter;

/**
 * Health check actor that monitors the health of other actors in the system
 * and provides health status information for monitoring and alerting.
 *
 * Equivalent to Scala's HealthCheckActor
 */
public class HealthCheckActor extends AbstractActor {

    private final LoggingAdapter log = Logging.getLogger(getContext().getSystem(), this);
    private final Map<String, ActorRef> monitoredActors = new HashMap<>();
    private final Map<String, HealthStatus> healthStatus = new HashMap<>();
    private Optional<Cancellable> healthCheckSchedule = Optional.empty();

    public static Props props() {
        return Props.create(HealthCheckActor.class);
    }

    @Override
    public void preStart() {
        log.info("HealthCheckActor started");
        // Schedule periodic health checks every 30 seconds
        healthCheckSchedule = Optional.of(
            getContext().getSystem().scheduler().schedule(
                Duration.ofSeconds(10),
                Duration.ofSeconds(30),
                getSelf(),
                new PerformHealthChecks(),
                getContext().getDispatcher(),
                getSelf()
            )
        );
    }

    @Override
    public void postStop() {
        healthCheckSchedule.ifPresent(Cancellable::cancel);
        log.info("HealthCheckActor stopped");
    }

    @Override
    public Receive createReceive() {
        return receiveBuilder()
            .match(RegisterActor.class, this::handleRegisterActor)
            .match(UnregisterActor.class, this::handleUnregisterActor)
            .match(PerformHealthChecks.class, msg -> performHealthChecks())
            .match(GetHealthStatus.class, msg ->
                getSender().tell(new SystemHealthStatus(new HashMap<>(healthStatus)), getSelf()))
            .match(GetActorHealth.class, this::handleGetActorHealth)
            .match(MarkActorUnhealthy.class, this::handleMarkActorUnhealthy)
            .match(MarkActorHealthy.class, this::handleMarkActorHealthy)
            .build();
    }

    private void handleRegisterActor(RegisterActor message) {
        monitoredActors.put(message.name(), message.actorRef());
        healthStatus.put(message.name(), HealthStatus.healthy());
        log.info("Registered actor for monitoring: {}", message.name());
        getSender().tell(new ActorRegistered(message.name()), getSelf());
    }

    private void handleUnregisterActor(UnregisterActor message) {
        monitoredActors.remove(message.name());
        healthStatus.remove(message.name());
        log.info("Unregistered actor from monitoring: {}", message.name());
        getSender().tell(new ActorUnregistered(message.name()), getSelf());
    }

    private void handleGetActorHealth(GetActorHealth message) {
        Optional<HealthStatus> status = Optional.ofNullable(healthStatus.get(message.name()));
        getSender().tell(new ActorHealthStatus(message.name(), status), getSelf());
    }

    private void handleMarkActorUnhealthy(MarkActorUnhealthy message) {
        healthStatus.put(message.name(), HealthStatus.unhealthy(message.reason()));
        log.warning("Actor {} marked as unhealthy: {}", message.name(), message.reason());
    }

    private void handleMarkActorHealthy(MarkActorHealthy message) {
        healthStatus.put(message.name(), HealthStatus.healthy());
        log.info("Actor {} marked as healthy", message.name());
    }

    private void performHealthChecks() {
        log.debug("Performing health checks");
        monitoredActors.forEach((name, actorRef) -> {
            // Simple liveness check by sending a ping message
            // In a real system, you would implement a more sophisticated health check
            try {
                // Actor is assumed healthy if it exists in our map
                HealthStatus currentStatus = healthStatus.get(name);
                if (currentStatus == null || !currentStatus.isHealthy()) {
                    healthStatus.put(name, HealthStatus.healthy());
                    log.debug("Health check passed for {}", name);
                }
            } catch (Exception ex) {
                healthStatus.put(name, HealthStatus.unhealthy("Health check failed: " + ex.getMessage()));
                log.warning("Health check failed for {}: {}", name, ex.getMessage());
            }
        });
    }

    // Messages
    public record RegisterActor(String name, ActorRef actorRef) {}
    public record UnregisterActor(String name) {}
    public record PerformHealthChecks() {}
    public record GetHealthStatus() {}
    public record GetActorHealth(String name) {}
    public record MarkActorUnhealthy(String name, String reason) {}
    public record MarkActorHealthy(String name) {}

    // Responses
    public record ActorRegistered(String name) {}
    public record ActorUnregistered(String name) {}
    public record SystemHealthStatus(Map<String, HealthStatus> status) {}
    public record ActorHealthStatus(String name, Optional<HealthStatus> status) {}

    // Health status types
    public sealed interface HealthStatus permits HealthStatus.Healthy, HealthStatus.Unhealthy {
        boolean isHealthy();

        static HealthStatus healthy() {
            return new Healthy();
        }

        static HealthStatus unhealthy(String reason) {
            return new Unhealthy(reason);
        }

        record Healthy() implements HealthStatus {
            @Override
            public boolean isHealthy() {
                return true;
            }
        }

        record Unhealthy(String reason) implements HealthStatus {
            @Override
            public boolean isHealthy() {
                return false;
            }
        }
    }
}

File: scala-example/src/main/scala/demo/actors/HealthCheckActor.scala
package demo.actors

import akka.actor.Actor
import akka.actor.ActorLogging
import akka.actor.ActorRef
import akka.actor.Cancellable
import akka.actor.Props

import scala.collection.mutable
import scala.concurrent.duration.*

/**
 * Health check actor that monitors the health of other actors in the system
 * and provides health status information for monitoring and alerting.
 */
class HealthCheckActor extends Actor with ActorLogging {

  import HealthCheckActor._
  import context.dispatcher

  private val monitoredActors = mutable.Map[String, ActorRef]()
  private val healthStatus = mutable.Map[String, HealthStatus]()
  private var healthCheckSchedule: Option[Cancellable] = None

  override def preStart(): Unit = {
    log.info("HealthCheckActor started")
    // Schedule periodic health checks every 30 seconds
    healthCheckSchedule = Some(
      context.system.scheduler.scheduleWithFixedDelay(
        initialDelay = 10.seconds,
        delay = 30.seconds,
        receiver = self,
        message = PerformHealthChecks
      )
    )
  }

  override def postStop(): Unit = {
    healthCheckSchedule.foreach(_.cancel())
    log.info("HealthCheckActor stopped")
  }

  override def receive: Receive = {
    case RegisterActor(name, actorRef) =>
      monitoredActors += name -> actorRef
      healthStatus += name -> HealthStatus.Healthy
      log.info(s"Registered actor for monitoring: $name")
      sender() ! ActorRegistered(name)

    case UnregisterActor(name) =>
      monitoredActors.remove(name)
      healthStatus.remove(name)
      log.info(s"Unregistered actor from monitoring: $name")
      sender() ! ActorUnregistered(name)

    case PerformHealthChecks =>
      performHealthChecks()

    case GetHealthStatus =>
      sender() ! SystemHealthStatus(healthStatus.toMap)

    case GetActorHealth(name) =>
      val status = healthStatus.get(name)
      sender() ! ActorHealthStatus(name, status)

    case MarkActorUnhealthy(name, reason) =>
      healthStatus += name -> HealthStatus.Unhealthy(reason)
      log.warning(s"Actor $name marked as unhealthy: $reason")

    case MarkActorHealthy(name) =>
      healthStatus += name -> HealthStatus.Healthy
      log.info(s"Actor $name marked as healthy")
  }

  private def performHealthChecks(): Unit = {
    log.debug("Performing health checks")
    monitoredActors.foreach { case (name, actorRef) =>
            // Simple liveness check by sending a ping message
      // In a real system, you would implement a more sophisticated health check
      try {
        // Actor is assumed healthy if it exists in our map
        healthStatus.get(name) match {
          case Some(HealthStatus.Healthy) => // Already healthy
          case _ =>
            healthStatus += name -> HealthStatus.Healthy
            log.debug(s"Health check passed for $name")
        }
      } catch {
        case ex: Exception =>
          healthStatus += name -> HealthStatus.Unhealthy(s"Health check failed: ${ex.getMessage}")
          log.warning(s"Health check failed for $name: ${ex.getMessage}")
      }
    }
  }
}

object HealthCheckActor {
  def props(): Props = Props(new HealthCheckActor)

  // Messages
  case class RegisterActor(name: String, actorRef: ActorRef)
  case class UnregisterActor(name: String)
  case object PerformHealthChecks
  case object GetHealthStatus
  case class GetActorHealth(name: String)
  case class MarkActorUnhealthy(name: String, reason: String)
  case class MarkActorHealthy(name: String)

  // Responses
  case class ActorRegistered(name: String)
  case class ActorUnregistered(name: String)
  case class SystemHealthStatus(status: Map[String, HealthStatus])
  case class ActorHealthStatus(name: String, status: Option[HealthStatus])

  // Health status types
  sealed trait HealthStatus
  object HealthStatus {
    case object Healthy extends HealthStatus
    case class Unhealthy(reason: String) extends HealthStatus
  }
}
